📖proposal for multi-az apiserver loadbalancer

sebltm opened the pull request:
<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->

- [ ] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

lentzi90 commented:
/ok-to-test

lentzi90 commented on docs/proposals/20250818-multi-az-apiserver-loadbalancer.md:
````diff
@@ -0,0 +1,298 @@
+  end
+```
+
+## Integration with External Global Server Load Balancing (GSLB)
````
This section about GSLB seems quite redundant to me. It doesn't really have anything to do with CAPO or the implementation of this feature. It just explains how it can fit in the larger ecosystem. I would prefer to reduce this to just a paragraph under the **Motivation** section.

lentzi90 commented on docs/proposals/20250818-multi-az-apiserver-loadbalancer.md:
````diff
@@ -0,0 +1,298 @@
+    allowCrossAZLoadBalancerMembers: true
+```
+
+Restrict access using allowed CIDRs
+```yaml
+spec:
+  apiServerLoadBalancer:
+    enabled: true
+    network:
+      id: 6c90b532-7ba0-418a-a276-5ae55060b5b0
+    availabilityZoneSubnets:
+      - availabilityZone: az1
+        subnet:
+          id: cad5a91a-36de-4388-823b-b0cc82cadfdc
+      - availabilityZone: az2
+        subnet:
+          id: e2407c18-c4e7-4d3d-befa-8eec5d8756f2
+    allowedCIDRs:
+      - 192.0.2.0/24
+      - 203.0.113.10
+```
````
This is not new in this proposal so I would leave it out.

lentzi90 commented on docs/proposals/20250818-multi-az-apiserver-loadbalancer.md:
```diff
@@ -0,0 +1,298 @@
+  - status.apiServerLoadBalancers is populated alongside legacy status until further cleanup.
+- Disabling multi-AZ:
+  - Remove the mapping; controller maintains single-LB behavior.
+  - Per-AZ LBs are not automatically deleted; operators may clean up unused resources.
```
Why would we not clean them up automatically?

lentzi90 commented on docs/proposals/20250818-multi-az-apiserver-loadbalancer.md:
```diff
@@ -0,0 +1,298 @@
+- DNS records remain out of scope for e2e.
+
+Test code locations:
+- [pkg/cloud/services/loadbalancer/](pkg/cloud/services/loadbalancer/)
+- [controllers/](controllers/)
+- [test/e2e/](test/e2e/)
```
Broken links

lentzi90 commented on docs/proposals/20250818-multi-az-apiserver-loadbalancer.md:
```diff
@@ -0,0 +1,298 @@
+- Keep the API additive with strong validation, clear events and documentation.
+- Preserve user-provided DNS endpoints; DNS record management remains out of scope.
+
+## Non-Goals
+- Managing or provisioning DNS records.
+- Provider-specific topologies such as ACTIVE_STANDBY across fault domains.
+- Service type LoadBalancer for worker Services.
```
It seems to me that automatic creation of the network and subnets is also a non-goal, right?

lentzi90 commented on docs/proposals/20250818-multi-az-apiserver-loadbalancer.md:
```diff
@@ -0,0 +1,298 @@
+## Design Overview
+
+### High-level behavior
+- When enabled and configured with an explicit mapping, CAPO reconciles one LoadBalancer per Availability Zone (AZ).
```
Perhaps add here that the operator creates the network and subnets?

lentzi90 reviewed (commented):
Interesting proposal! It will need some polishing though.
I think all the links are broken, so that would perhaps be the first thing to fix.

sebltm commented:
/remove-lifecycle stale

lentzi90 reviewed (commented):
Looking pretty good!
Could you squash the commits?

I would also be interested in other maintainers opinion on this.
/cc @mandre @stephenfin

sebltm committed:
feat: proposal for multi-az apiserver loadbalancer

Add a design proposal for first-class Multi-AZ support for the
Kubernetes control plane LoadBalancer in CAPO. The feature reconciles
one Octavia LoadBalancer per Availability Zone via an explicit
AZ-to-Subnet mapping, registers control plane nodes with the LB in
their AZ by default, and relies on external DNS multi-value A records
for client-side failover.

lentzi90 commented:
@nikParasyr have you seen this proposal? I would be interested in your thoughts, especially when considering v1beta2 changes we want to make (e.g. #2899).

nikParasyr reviewed (commented):
I have 3 main "concerns" here:
1. The change to `OSC.status` will be a breaking change as we will need to go from `status.apiServerLoadBalancer` to `status.apiServerLoadBalancer[]`. We could introduce is at as a new field and deprecate the other but its something to keep in mind.
2. There are some fields of the current api that are not covered by the proposal. For example `.spec.apiServerFloatingIP`. I suspect that being able to predefine the floating ip per LB/AZ would be nice as it would allow users to create the FIPs & records beforehand (or through a lifecycleHook) and let the cluster roll out nicely. This would affect the current v1beta2 proposed changes as we will probably have to opt for [option 2](https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/2899#issuecomment-3777687418) and might even have to change the managedLoadbalancer to a list.
3. The request in https://github.com/kubernetes-sigs/cluster-api-provider-openstack/issues/2999 has to be incorporated in this proposal i think. It doesnt feel usable to me (but might be missing something as openstack deployments vary) if we can create LBs per AZ on different subnets but cannot put the control-plane nodes on these subnets.

My other concern is that this will be tricky to test in e2e, but lets see

sebltm committed:
fix: attempt to address comments

lentzi90 reviewed (commented):
I am happy with this now. Let's see what others think.
/approve

lentzi90 commented:
/cc @bnallapeta

sebltm commented:
@lentzi90 anything I can do to help move this one forward?

lentzi90 commented:
Not much we can do other than trying to ask for more reviews. I'll put it in the slack channel and mention it in the office hours :slightly_smiling_face:

nikParasyr reviewed (approved):
/lgtm

My concerns in regards with e2e tests and "partial implementation" remain, this is why i was sceptical approving this. But i guess approving this would be the way to start tackling them.

lentzi90 commented:
Agreed!
/hold cancel
